### PR TITLE
Add take() aggregation function in PPL

### PR DESCRIPTION
**Pull request by joshuali925**
### Description
`take(field)` returns some documents of a field per each `stats` response. It is implemented using top hits metrics aggregation when pushed down to opensearch. In memory it works similar to `head` command.

A sample use case would be aggregating logs by a category, `take` allows user to see a sample log per each category: `stats take(log_string, 1) by category`.

This returns a list of patterns with the count and one sample message. 

Currently `take` only supports text fields. it's not very useful on other types, and array implementation might need to be improved. Support for other types can be extended in the future if needed.
 
### Issues Resolved
Linking issue: https://github.com/opensearch-project/observability/issues/1172
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: docs/user/ppl/cmd/stats.rst [hunk 1]
```diff
@@ -238,6 +238,28 @@ Example::
...
+----------
+
+Description
+>>>>>>>>>>>
+
+Usage: TAKE(field [, size [, from]]). Return original values of a field.
+
+* field: mandatory. The field must be a text field.
+* size: optional number. The number of values should be returned. Default is 10.
+* from: optional number. The number of values should be skipped. Default is 0.
```
Comment (joshuali925) [inline docs/user/ppl/cmd/stats.rst:251]: just realized `from` doesn't make much sense without sort, will remove it for now

**Comment by Yury-Fridlyand** [comment]
> Currently take only supports text fields
Do you have a test for non-text fields?
A test on field  which doesn't exit?

**Comment by joshuali925** [comment]
> > Currently take only supports text fields
> 
> Do you have a test for non-text fields? A test on field which doesn't exit?

That should be responsibilities of function resolver and expression analyzer? i didn't modify those logic, i think UT already covers those. If it happens error will say
```
take function expected {[STRING,INTEGER]}, but get [LONG,INTEGER]
```
or 
```
can't resolve Symbol(namespace=FIELD_NAME, name=nonexist) in type env
```

File: opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java [hunk 2]
```diff
@@ -214,7 +214,13 @@ private ExprValue parseStruct(Content content, String prefix) {
    */
   private ExprValue parseArray(Content content, String prefix) {
     List<ExprValue> result = new ArrayList<>();
-    content.array().forEachRemaining(v -> result.add(parse(v, prefix, Optional.of(STRUCT))));
+    content.array().forEachRemaining(v -> {
+      if (v.isString()) {
```
Comment (vamsi-amazon) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java:218]: Why does this required a special case. Isn't string type handled by parse function.

File: docs/user/ppl/cmd/stats.rst [hunk 3]
```diff
@@ -238,6 +238,27 @@ Example::
...
+>>>>>>>>>>>
+
+Usage: TAKE(field [, size]). Return original values of a field.
+
+* field: mandatory. The field must be a text field.
+* size: optional integer. The number of values should be returned. Default is 10.
+
+Example::
+
+    os> source=accounts | stats take(firstname);
```
Comment (vamsi-amazon) [inline docs/user/ppl/cmd/stats.rst:254]: Would you consider adding more examples with groupBy, size and may be patterns if possible?
or add a complex Example which covers all.
will you be including ITs in this PR or other PR?

**Comment by vamsi-amazon** [comment]
When we push down to OS, what is the default sort applied? Will it be consistent with our in-memory implementation?

File: opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java (hunk 2 above)
Comment (joshuali925) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java:218]: by default it's parsing an array of structs, not an array of strings

File: docs/user/ppl/cmd/stats.rst (hunk 3 above)
Comment (joshuali925) [reply docs/user/ppl/cmd/stats.rst:254]: will add one with groupBy and ITs. I don't know if we should put `patterns` here since this is the stats doc

**Comment by joshuali925** [comment]
> When we push down to OS, what is the default sort applied? Will it be consistent with our in-memory implementation?

`take` command doesn't guarantee on order

File: ppl/src/main/java/org/opensearch/sql/ppl/parser/AstExpressionBuilder.java [hunk 4]
```diff
@@ -214,6 +214,17 @@ public UnresolvedExpression visitPercentileAggFunction(PercentileAggFunctionCont
         Collections.singletonList(new Argument("rank", (Literal) visit(ctx.value))));
   }
 
+  @Override
+  public UnresolvedExpression visitTakeAggFunctionCall(
+      OpenSearchPPLParser.TakeAggFunctionCallContext ctx) {
+    ImmutableList.Builder<UnresolvedExpression> builder = ImmutableList.builder();
+    builder.add(new UnresolvedArgument("size",
+        ctx.takeAggFunction().size != null ? visit(ctx.takeAggFunction().size) :
+            AstDSL.integerLiteral(10)));
```
[the remark this answers is not in the document]
Comment (joshuali925) [reply ppl/src/main/java/org/opensearch/sql/ppl/parser/AstExpressionBuilder.java]: sure, moved to static const

File: docs/user/ppl/cmd/stats.rst (hunk 3 above)
Comment (joshuali925) [reply docs/user/ppl/cmd/stats.rst:254]: added example 10 https://github.com/opensearch-project/sql/blob/2ad0f5eab9609cc6ced106a06ce7303ac396d425/docs/user/ppl/cmd/stats.rst#example-10-calculate-the-count-and-get-email-list-by-a-gender-and-span

**Review by vamsi-amazon — DISMISSED** [review]

File: opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java (hunk 2 above)
Comment (penghuo) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java:218]: add comments?

File: docs/user/ppl/cmd/stats.rst [hunk 5]
```diff
@@ -238,6 +238,27 @@ Example::
     | 2.8613807855648994 |
     +--------------------+
 
+TAKE
+----------
+
+Description
+>>>>>>>>>>>
+
+Usage: TAKE(field [, size]). Return original values of a field.
```
Comment (penghuo) [inline docs/user/ppl/cmd/stats.rst:247]: could you add more detail? e.g. how does the original values picked, random? or ordered?

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/aggregation/dsl/MetricAggregationBuilder.java [hunk 6]
```diff
@@ -194,6 +204,27 @@ private Pair<AggregationBuilder, MetricParser> make(CardinalityAggregationBuilde
...
+   */
+  private Pair<AggregationBuilder, MetricParser> make(TopHitsAggregationBuilder builder,
+                                                      Expression expression,
+                                                      Expression size,
+                                                      Expression condition,
+                                                      String name,
+                                                      MetricParser parser) {
+    String fieldName = ((ReferenceExpression) expression).getAttr();
+    builder.fetchSource(fieldName, null);
+    builder.size(size.valueOf(null).integerValue());
```
Comment (penghuo) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/aggregation/dsl/MetricAggregationBuilder.java:218]: i saw a lot of place use xxx.valueOf(null) seems Expression should define `ExprValue valueOf()` interface, and only LiteralExpression has implementation.

Comment (penghuo) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/aggregation/dsl/MetricAggregationBuilder.java:218]: we can create the issue and make the change in following PR.

File: docs/user/ppl/cmd/stats.rst [hunk 7]
```diff
@@ -381,3 +402,20 @@ PPL query::
...
+The example gets the count of age by the interval of 10 years and group by gender, additionally for each row get a list of at most 5 emails.
+
+PPL query::
+
+    os> source=accounts | stats count() as cnt, take(email, 5) by span(age, 5) as age_span, gender
+    fetched rows / total rows = 3/3
+    +-------+--------------------------------------------+------------+----------+
+    | cnt   | take(email, 5)                             | age_span   | gender   |
+    |-------+--------------------------------------------+------------+----------|
+    | 1     | []                                         | 25         | F        |
```
[the remark this answers is not in the document]
Comment (joshuali925) [reply docs/user/ppl/cmd/stats.rst:418]: yes, json array

File: core/src/main/java/org/opensearch/sql/ast/dsl/AstDSL.java [hunk 8]
```diff
@@ -199,6 +199,10 @@ public static Literal stringLiteral(String value) {
     return literal(value, DataType.STRING);
   }
 
+  public static Literal integerLiteral(Integer value) {
```
[the remark this answers is not in the document]
Comment (joshuali925) [reply core/src/main/java/org/opensearch/sql/ast/dsl/AstDSL.java:202]: missed intLiteral, will remove this one..

File: docs/user/ppl/cmd/stats.rst (hunk 5 above)
Comment (joshuali925) [reply docs/user/ppl/cmd/stats.rst:247]: I'll add `Take command does not guarantee on order of values.`, let me know if i should mention top_hits here.

if it's pushed down it's top_hits order, in memory it is same as `head`

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/aggregation/dsl/MetricAggregationBuilder.java (hunk 6 above)
Comment (joshuali925) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/aggregation/dsl/MetricAggregationBuilder.java:218]: https://github.com/opensearch-project/sql/issues/976

**Review by MaxKsyunz — APPROVED** [review]

**Review by penghuo — APPROVED** [review]

**Pull request merged**